doc/compiler/aot/Diagnostics.md

```diff
@@ -0,0 +1,396 @@
+```
+Type                                File                             Line  Offset(M) Offset(PC) Node
+TR_InlinedVirtualMethodWithNopGuard (34) /root/hostdir/openj9-openjdk-jdk11/openj9/runtime/compiler/codegen/J9CodeGenerator.cpp  2774      007e       003e 0000000000000000
+TargetAddress1:00007F470F23B980,  TargetAddress2:0000000000000000
+
+```
+This section contains information about the abstract notion of the
+relocation record, namely the `TR::ExternalRelocationRecord`. Objects
+of this type are iterated over, and used to generate the actual
+relocation records that are stored into the SCC, described by the
+`TR_RelocationRecordHelperAddressBinaryTemplate` family of structs.
+
+Here, `TR_InlinedVirtualMethodWithNopGuard` is the relocation type,
+specifically the `TR_ExternalRelocationTargetKind` enum; `34` is the
+value of the enum. Also printed is the file and line number where this
+External Relocation was generated. Next follows two offsets, the first
+is the offset from the start of the allocated memory for this method,
+and the second is the offset from the startPC. This is followed by the
+`TR::Node` that is associated with this relocation record; this is used
+to connect the relocation record with the IL. In the case of validation
+or inlined site records, this may be NULL; however for the
+former the node does not matter and for the latter, the inlined site
+index can be determined from the next section below. Finally, the two
+values passed into the constructor of the `TR::ExternalRelocationRecord`
+is printed out; these values are used to generate the values in the
+actual relocation record that gets stored into the SCC.
+
+#### Binary Data Info
+```
+Address           Size Type                           Width EIP Index Offsets
+         eb4b0a0  50   TR_InlinedVirtualMethodWithNopGuard (34)2     Abs       007e
+Inlined Method: Inlined site index = 0, Constant pool = dd4ffa18, cpIndex = 1c, romClassOffsetInSharedCache=0000000000042F70, destinationAddress = 00007F4714DE7100
+```
+This section contains information about the actual relocation record
+that is stored into the SCC. This data is described by the
+`TR_RelocationRecordHelperAddressBinaryTemplate` family of structs.
+These structs are variable length; they are followed by a variable
+number of 2 or 4 byte offsets.
+
+Here, `eb4b0a0` is the address of start of one the Binary Template
+structs; it is only valid during the compilation as the entire buffer
+is stored into the SCC. Next is the size of the entire variable length
+record, followed by the `TR_ExternalRelocationTargetKind` associated
+with this record. Then comes the size of the offsets, and whether or
+not the relocation is relative to the instruction pointer. Finally the
+list of offsets in the code the relocation record applies to; the
+offset here is the `Offset (M)` from above. Also printed out on the
+next line are the various fields in the Binary Template; this will vary
+depending on the relocation kind.
```

Comment:
Maybe this paragraph could also be organized in bulletin points?

Reply:
Update in force push.